tests/conftest.py

```diff
@@ -2,29 +2,254 @@
 
 from __future__ import annotations
 
-from typing import AsyncIterator
+import asyncio
+from copy import copy
+from pathlib import Path
+from typing import AsyncIterator, Iterator
 
+import pytest
 import pytest_asyncio
+import structlog
 from asgi_lifespan import LifespanManager
 from fastapi import FastAPI
 from httpx import AsyncClient
+from safir import logging
+from safir.kubernetes import initialize_kubernetes
+from structlog.stdlib import BoundLogger
 
-from jupyterlabcontroller import main
+from jupyterlabcontroller.config import Configuration
+from jupyterlabcontroller.dependencies.config import configuration_dependency
+from jupyterlabcontroller.dependencies.prepull import (
+    prepuller_manager_dependency,
+)
+from jupyterlabcontroller.dependencies.storage import (
+    docker_storage_dependency,
+    gafaelfawr_storage_dependency,
+    k8s_storage_dependency,
+)
+from jupyterlabcontroller.main import create_app
+from jupyterlabcontroller.models.context import Context
+from jupyterlabcontroller.models.domain.docker import DockerCredentialsMap
+from jupyterlabcontroller.models.domain.storage import StorageClientBundle
+from jupyterlabcontroller.models.domain.usermap import UserMap
+from jupyterlabcontroller.models.v1.lab import UserInfo
+from jupyterlabcontroller.services.prepuller import PrepullerManager
+from jupyterlabcontroller.storage.docker import DockerStorageClient
+from jupyterlabcontroller.storage.gafaelfawr import GafaelfawrStorageClient
+from jupyterlabcontroller.storage.k8s import K8sStorageClient
+
+from .settings import TestObjectFactory, test_object_factory
+from .support.mockdocker import MockDockerStorageClient
+from .support.mockgafaelfawr import MockGafaelfawrStorageClient
+from .support.mockk8s import MockK8sStorageClient
+
+_here = Path(__file__).parent
+
+CONFIG_DIR = Path(_here / "configs/standard")
+
+"""Change the test application configuration to point at a file that
+replaces the YAML that would usually be mounted into the container at
+``/etc/nublado/config.yaml``.  For testing and standalone purposes, if
+the filename is not the standard location, we expect the Docker
+credentials (if any) to be in ``docker_config.json`` in the same
+directory as ``config.yaml``, and we expect objects used in testing to
+be in ``test_objects.json`` in that directory.
+"""
+
+
+@pytest.fixture
+def logger() -> BoundLogger:
+    return structlog.get_logger(logging.logger_name)
+
+
+@pytest.fixture
+def obj_factory() -> TestObjectFactory:
+    filename = str(CONFIG_DIR / "test_objects.json")
+    test_object_factory.initialize_from_file(filename)
+    return test_object_factory
+
+
+@pytest.fixture
+def http_client() -> AsyncClient:
+    return AsyncClient(follow_redirects=True)
```

Comment:
You generally shouldn't need a fixture for this; you can just call `http_client_dependency()` when you need it.

Reply:
Done (but there's a lot of fixture work to be done to adopt the process_context/request_context + factory pattern)